network/discovery/src/file.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::DiscoveryError;
+use diem_config::config::PeerSet;
+use diem_time_service::{Interval, TimeService, TimeServiceTrait};
+use futures::Stream;
+use std::{
+    path::{Path, PathBuf},
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+pub struct FileStream {
+    file_path: PathBuf,
+    interval: Pin<Box<Interval>>,
+}
+
+impl FileStream {
+    pub(crate) fn new(
+        file_path: &Path,
+        interval_duration: Duration,
+        time_service: TimeService,
+    ) -> Self {
+        FileStream {
+            file_path: file_path.to_path_buf(),
+            interval: Box::pin(time_service.interval(interval_duration)),
+        }
+    }
+}
+
+impl Stream for FileStream {
+    type Item = Result<PeerSet, DiscoveryError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Wait for delay, or add the delay for next call
+        futures::ready!(self.interval.as_mut().poll_next(cx));
+
+        Poll::Ready(Some(match load_file(self.file_path.as_path()) {
```

Comment:
feels like we should only be reading the file if it has been updated, it doesn't look like tokio provides any event apis, but we can just read the metadata: https://docs.rs/tokio/0.1.22/tokio/fs/struct.File.html#method.metadata

maybe make a follow up task?

Reply:
There is a [notify](https://docs.rs/notify/5.0.0-pre.10/notify/index.html) crate for this, but given our reluctance to import new crates I skipped this for now.

It can be included in the other followup task on reading with tokio

Reply:
Issue for addressing both https://github.com/diem/diem/issues/8560